src/common.rs

```diff
@@ -124,3 +124,69 @@ pub fn get_widgets() -> color_eyre::eyre::Result<
     }
     Ok(widgets)
 }
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct PredecessorIdFunctionArgs {
+    predecessor_id: near_primitives::types::AccountId,
+    key: String,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct PublicKeyFunctionArgs {
+    public_key: near_crypto::PublicKey,
+    key: String,
+}
+
+pub async fn is_write_permission_granted(
+    config: near_cli_rs::config::Config,
+    network_config: near_cli_rs::network_for_transaction::NetworkForTransactionArgs,
+    near_social_account_id: near_primitives::types::AccountId,
+    predecessor_id: Option<near_primitives::types::AccountId>,
+    public_key: Option<near_crypto::PublicKey>,
```

Comment:
Instead of two Option parameters, create a new enum `PermissionKey`, and then you will be able to define a single struct for the `is_write_permission_granted` function arguments:

```rust
#[derive(...)]
enum PermissionKey {
    PredecessorId(near_primitives::types::AccountId),
    PublicKey(near_crypto::PublicKey),
}

impl From<near_primitives::types::AccountId> for PermissionKey {
    fn from(predecessor_id: near_primitives::types::AccountId) -> Self {
        Self::PredecessorId(predecessor_id)
    }
}

impl From<near_crypto::PublicKey> for PermissionKey {
    fn from(public_key: near_crypto::PublicKey) -> Self {
        Self::PublicKey(public_key)
    }
}

#[derive(...)]
struct IsWritePermissionGrantedInputArgs {
    key: String,
    #[serde(flatten)]
    permission_key: PermissionKey,
}

pub async fn<P: Into<PermissionKey>> is_write_permission_granted(
    ...
    permission_key: P,
    key: String
) {
    let function_args = serde_json::to_string(&IsWritePermissionGrantedInputArgs {
            key,
            permission_key: permission_key.into(),
        })
        .wrap_err("Internal error: could not serialize `is_write_permission_granted` input args")?;
   ...
}

// and then you will be able to call it just like this:
is_write_permission_granted(..., signer_account_id, key).await
is_write_permission_granted(..., signer_public_key, key).await
```

Note, `#[serde(flatten)]` is important there.

Reply:
done